### ISSUE TITLE
todo: Refactor `PERMIT_DOCKER` support

### ISSUE THREAD

**Issue by polarathene** [docker-mailserver/docker-mailserver#3478]
### Subject

I would like to contribute to the project

### Description

Tasks below in the TLDR section are available for anyone willing to tackle them. None of it is urgent, mostly improvements in better UX / compatibility.

---

## Current state overview

Technically this is dev docs, should anyone maintain this feature in future.

[`PERMIT_DOCKER` ENV](https://docker-mailserver.github.io/docker-mailserver/v12.1/config/environment/#permit_docker) modifies Postfix [`mynetworks`](https://www.postfix.org/postconf.5.html#mynetworks) which is [focused specifically on what networks are trusted to bypass restrictions](https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/postfix/main.cf#L52-L62).

![image](https://github.com/docker-mailserver/docker-mailserver/assets/5098581/21cf192f-99be-4fc5-947a-c76574dfff35)

The [`NETWORK_INTERFACE` ENV](https://docker-mailserver.github.io/docker-mailserver/latest/config/environment/#network_interface) only seems to be used for supporting the `PERMIT_DOCKER` ENV (_which it queries for the specified interface (default `eth0`) and retrieves the IPv4 address only. The docs advise setting another ENV to configure Postfix to only listen on IPv4 for compatibility_):

https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/startup/setup.d/networking.sh#L16-L27

Above is extracted/used below for the different `PERMIT_DOCKER` settings:

- `CONTAINER_IP` is extracted for `PERMIT_DOCKER=container` (_IP assigned from the interface to the DMS container_):

   https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/startup/setup.d/networking.sh#L57-L59

- `CONTAINER_NETWORK` is extracted (_as a CIDR `/16` subnet (last two octets) of the `CONTAINER_IP` to belong to_) for `PERMIT_DOCKER=host`. While intended for host mode networking, `/16` may not always be an appropriate size? 🤷‍♂️ :

   https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/startup/setup.d/networking.sh#L61-L63

- `PERMIT_DOCKER=network` uses a standard docker IPv4 subnet, but assumes that the container is assigned to it (_likely not compatible with the compose bridge network used_):

   https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/startup/setup.d/networking.sh#L65-L67

   **NOTE:** The IPv4 subnet pool used by docker also [varies in subnet size](https://github.com/moby/moby/discussions/46033#discussioncomment-6510204).

- While `PERMIT_DOCKER=connected-networks` is the extended variant, iterating through all IPv4 networks visible (_technically it [filters to only `veth` type](https://serverfault.com/questions/1019363/using-ip-address-show-type-to-display-physical-network-interface)_) in the container and whitelists these as subnets of size `/16`:

   https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/startup/setup.d/networking.sh#L25-L27

   https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/startup/setup.d/networking.sh#L50-L55

   https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/helpers/network.sh#L19-L23

Those all then get whitelisted by Postfix (`main.cf:mynetworks`) and the OpenDKIM/OpenDMARC services (_but not others services we have added since? May introduce some unexpected failures?_ 🤷‍♂️ ):

https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/startup/setup.d/networking.sh#L34-L43

---

## TL;DR (plus few other related concerns)

This ENV / feature was implemented originally in 2016: https://github.com/docker-mailserver/docker-mailserver/pull/270

TODO:
- `_sanitize_ipv4_to_subnet_cidr()` could probably be used to get the actual CIDR mask for the network?
- `PERMIT_DOCKER=container` shouldn't require `/32`? Same for `::1` with `/128`? (_In Postfix `main.cf:mynetworks`_)
- IPv6 addresses could avoid the `[]` notation in `mynetworks` by referring to a file instead, [CIDR table](https://www.postfix.org/cidr_table.5.html). This would [avoid the `postconf` command](https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/startup/setup.d/networking.sh#L39-L40) (_except for `PERMIT_DOCKER=none`_), which is simpler (_append a line, no `mtime` workaround call_).
- This is probably not container restart friendly as we don't presently reset the files updated AFAIK?:

   https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/startup/setup.d/networking.sh#L40-L42
- Possibly a bit more tricky, is adding IPv6 support as the current script support is IPv4 focused. Python, Rust, or similar may make that easier than messing around with bash 😅 
- Instead of the default `eth0`, getting the interface of the default route [might work just as well when queried?](https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/startup/setup.d/networking.sh#L16), perhaps even with Podman which [our docs mention `tap0` name](https://docker-mailserver.github.io/docker-mailserver/latest/config/advanced/podman/#use-the-slip4netns-network-driver).
- [These Rootless (Podman) docs](https://docker-mailserver.github.io/docker-mailserver/latest/config/advanced/podman/#security-in-rootless-mode) should probably be linked to from the `PERMIT_DOCKER` ENV docs, as that's quite the caveat to be aware of when configuring this ENV. Additionally, those docs appear outdated as `PERMIT_DOCKER=none` is the [current default and `mailserver.env` reflects that](https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/mailserver.env#L69).
- `PERMIT_DOCKER=network` + `PERMIT_DOCKER=host` would become roughly the same logic if they just resolve the subnet the container IP belongs to (`NETWORK_INTERFACE`). The name `PERMIT_DOCKER` could probably benefit with a more apt name `TRUSTED_NETWORKS`.
- Do note that until Docker addresses some issues with `userland-proxy` (iptables rules + proxy), the current `daemon.json` default they default `userland-proxy: true` which especially for setups not configuring IPv6 can also introduce an Open Relay by enabling clients to masquerade as the gateway of a Docker network (_that belongs to a trusted subnet via `PERMIT_DOCKER`, thus bypassing restrictions_).

Postfix [`mynetworks` also has our `main.cf` include `fe80::/64`](https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/postfix/main.cf#L17C47-L17C47) as a default. AFAIK the actual prefix would be `fe80::/10`, while the practical assignment is a `/64` the full network prefix (first 64-bits) is created and thus won't match `fe80::/64` which would have 52-bits of 0 padding (wrong network). You probably don't want `fe80::/10` to be allowed by default though as I believe that'd permit any containers on the same network to be trusted (_not a good default_). So it could probably be dropped, and maybe documented until someone raises a case for it to be default in DMS? It was part of the original PR, but they didn't explain why IPv6 link-local address was important for Postfix (_[it's possibly valid](https://mailing.postfix.users.narkive.com/OYLwXCj9/mynetworks-support-for-ipv6-link-local-fe80-hosts), but probably unlikely these days?_).

**Comment by polarathene** [comment]
## TODO:

Probably should document the way trust works, with `permit_mynetworks` (_`PERMIT_DOCKER` ENV_) and the often better choice `permit_sasl_authenticated` (_authenticated submission via port 587 / 465 with credentials to a DMS account_).
- I [replied here](https://github.com/docker-mailserver/docker-mailserver/issues/3556#issuecomment-1740382392) describing the two and the present risks with Dockers default networking that `PERMIT_DOCKER=network` may expose unintentionally.
- Also [another comment](https://github.com/docker-mailserver/docker-mailserver/issues/1405#issuecomment-1709938022) of mine from earlier answers some questions about Open Relay concerns and how a user can try verify the risk if they insist on trusting a subnet entirely.

**Comment by georglauterbach** [comment]
I'd like to propose a new approach:

1. Deprecate `PERMIT_DOCKER` and introduce `LOCAL_NETWORKS_IPV4` and `LOCAL_NETWORKS_IPV6`.
2. Use these ENV's values to write the appropriate configuration to all of our services (Postfix, Rspamd, OpenDKIM, OpenDMARC, you name it).

---

```env
# This variable should contain all of your local IPv4 networks (in CIDR notation). These
# networks are handled in a special way in many services. We insert the networks into
#
# Postfix'     `mynetworks`, "[a] list of 'trusted' remote SMTP clients that have more privileges than 'strangers'";
# Rspamd's     `local_networks`, where "certain checks are skipped [...] (e.g. SPF checks)" for entries in this list;
# OpenDMARC's  `/etc/opendmarc/ignore.hosts`, where hosts are ignored if they are on this list;
# OpenDKIM's   `/etc/opendkim/TrustedHosts` so OpenDKIM knows which hosts' e-mails it can sign.
#
# NOTE: If you want the IP(s) that used for the default route (probably on-link) (likely your public IPs), add "GATEWAYS" to the list (e.g.: LOCAL_NETWORKS_IPV4="127.0.0.1/8 192.168.0.0/16 GATEWAYS").
LOCAL_NETWORKS_IPV4="127.0.0.1/8 GATEWAYS"

# Like LOCAL_NETWORKS_IPV4 but for IPv6 addresses.
LOCAL_NETWORKS_IPV6.
```

```bash
#! /usr/bin/env bash

LOCAL_NETWORKS_IPV4="127.0.0.1/8 GATEWAY 192.168.0.0/78 10.0.10.3/24"

# Checks whether an IPv4 address is valid.
function _is_ipv4_address() {
  local NETWORK ADDRESS_PARTS SUFFIX
  NETWORK=${1:?IPv4 address must be provided}

  # Check coarsely whether the format `OCTET.OCTET.OCTET.OCTET/SUBNET` is being used
  [[ ${NETWORK} =~ ^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}/[0-9]{1,2}$ ]] || return 1

  # Acquire the individual octets of the address and check them
  mapfile -t -d '.' ADDRESS_PARTS < <(cut -z -d '/' -f 1 <<< "${NETWORK}")
  for OCTET in "${ADDRESS_PARTS[@]}"; do
    [[ ${OCTET} -ge 0 ]] && [[ ${OCTET} -le 255 ]] || return 1
  done

  # Lastly, check whether the suffix is valid
  SUFFIX=$(cut -d '/' -f 2 <<< "${NETWORK}")
  [[ ${SUFFIX} -ge 0 ]] && [[ ${SUFFIX} -le 32 ]] || return 1
}

function _mask_ip_digit() {
  if [[ ${1} -ge 8 ]]; then
    MASK=255
  elif [[ ${1} -le 0 ]]; then
    MASK=0
  else
    VALUES=(0 128 192 224 240 248 252 254 255)
    MASK=${VALUES[${1}]}
  fi

  local DVAL=${2}
  ((DVAL&=MASK))

  echo "${DVAL}"
}

function _sanitize_ipv4_to_subnet_cidr() {
  local DIGIT_PREFIX_LENGTH="${1#*/}"

  declare -a MASKED_DIGITS DIGITS
  IFS='.' read -r -a DIGITS < <(echo "${1%%/*}")

  for ((i = 0 ; i < 4 ; i++)); do
    MASKED_DIGITS[i]=$(_mask_ip_digit "${DIGIT_PREFIX_LENGTH}" "${DIGITS[i]}")
    DIGIT_PREFIX_LENGTH=$((DIGIT_PREFIX_LENGTH - 8))
  done

  echo "${MASKED_DIGITS[0]}.${MASKED_DIGITS[1]}.${MASKED_DIGITS[2]}.${MASKED_DIGITS[3]}/${1#*/}"
}

function _setup_local_networks() {
  function _add_to_postfix_mynetworks() {
    :
  }

  function _add_to_rspamd_local_networks() {
    :
  }

  function _add_to_opendkim_trusted_hosts() {
    :
  }

  function _add_to_opendmarc_ignore_hosts() {
    :
  }

  _log 'debug' "Adding local IPv4 networks '${LOCAL_NETWORKS_IPV4}' to services now"

  # Convert the string into an array we can work with
  # Note: Do not remove the whitespace after `%s`; ref: https://stackoverflow.com/questions/12916352/shell-script-read-missing-last-line
  IFS=' ' mapfile -t -d ' ' NETWORKS < <(printf '%s ' "${LOCAL_NETWORKS_IPV4}")

  local NETWORK
  for NETWORK in "${NETWORKS[@]}"; do
    _log 'trace' "Raw network to add is '${NETWORK}'"

    if [[ ${NETWORK} == 'GATEWAY' ]]; then
      NETWORK=$(ip route get default | /usr/bin/grep 'dev' | cut -d ' ' -f 4)
      NETWORK=$(ip -4 addr show dev "${NETWORK}" | grep 'inet' | awk '{$1=$1;print}' | cut -d ' ' -f 2)
      if [[ ${NETWORKS[*]} == *${NETWORK}* ]]; then
        _log 'warning' "Gateway network '${NETWORK}' already listed in local networks ('${NETWORKS[*]}') - skipping"
        continue
      fi
    fi

    if ! _is_ipv4_address "${NETWORK}"; then
      _log 'warning' "'${NETWORK}' is not a valid description of an IPv4 network in CIDR notation - skipping"
      continue
    fi

    NETWORK=$(_sanitize_ipv4_to_subnet_cidr "${NETWORK}")
    _log 'trace' "Adding network '${NETWORK}'"

    _add_to_postfix_mynetworks
    _add_to_rspamd_local_networks
    _add_to_postfix_mynetworks
    _add_to_postfix_mynetworks
  done
}

_setup_local_networks
```

**Comment by 53c70r** [comment]
I really appreciate the functionality to add custom mynetworks. <3

**Comment by georglauterbach** [comment]
PS: We can also easily extend the `GATEWAY`-approach with things like `DOCKER` to add the default Docker network, etc.

**Comment by luckman212** [comment]
@53c70r did you figure out how to add custom (ad-hoc) subnets or IPs to `mynetworks`?

**Comment by hanscees** [comment]
> I'd like to propose a new approach:
> 
> 1. Deprecate `PERMIT_DOCKER` and introduce `LOCAL_NETWORKS_IPV4` and `LOCAL_NETWORKS_IPV6`.
> 2. Use these ENV's values to write the appropriate configuration to all of our services (Postfix, Rspamd, OpenDKIM, OpenDMARC, you name it).
> 
> Entry in `mailserver.env`:
> 
> ```dotenv
> # This variable should contain all of your local IPv4 networks (in CIDR notation). These
> # networks are handled in a special way in many services. We insert the networks into
> #
> # Postfix'     `mynetworks`, "[a] list of 'trusted' remote SMTP clients that have more privileges than 'strangers'";
> # Rspamd's     `local_networks`, where "certain checks are skipped [...] (e.g. SPF checks)" for entries in this list;
> # OpenDMARC's  `/etc/opendmarc/ignore.hosts`, where hosts are ignored if they are on this list;
> # OpenDKIM's   `/etc/opendkim/TrustedHosts` so OpenDKIM knows which hosts' e-mails it can sign.
> #
> # ### Special Values
> #
> # 1. If you want the IP(s) that used for the default route (probably on-link) (likely your public IPs), add "GATEWAYS" to the list (e.g.: LOCAL_NETWORKS_IPV4="127.0.0.1/8 192.168.0.0/16 GATEWAYS").
> LOCAL_NETWORKS_IPV4="127.0.0.1/8 GATEWAYS"
> 
> # Like LOCAL_NETWORKS_IPV4 but for IPv6 addresses.
> LOCAL_NETWORKS_IPV6.
> ```
> 
> Note how this is pretty extensible with the special-values field :)
> 
> Updated `network.sh`:
> 
> ```shell
> # Split an IPv4 address into its individual octets and create an array.
> #
> # @param ${1} = name of the array that is created
> # @param ${2} = IPv4 address
> #
> # #### Attention
> #
> # This function assumes ${2} is valid to begin with!
> function _split_ipv4_into_octets() {
>   local ARRAY_NAME IPV4_NETWORK
>   ARRAY_NAME=${1:?Array name must be provided to _split_ipv4_into_octets}
>   IPV4_NETWORK=${2:?IPv4 address must be provided to _split_ipv4_into_octets}
> 
>   mapfile -t -d '.' "${ARRAY_NAME}" <<< "${IPV4_NETWORK%%/*}"
> }
> 
> # Checks whether an IPv4 address is valid.
> #
> # @param ${1} = IPv4 address
> function _is_valid_ipv4_address() {
>   local ADDRESS
>   ADDRESS=${1:?IPv4 address must be provided to _is_valid_ipv4_address}
>   [[ ${ADDRESS} =~ ^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$ ]] || return 1
> 
>   # Acquire the individual octets of the address and check them
>   _split_ipv4_into_octets 'ADDRESS_PARTS' "${ADDRESS}"
>   for OCTET in "${ADDRESS_PARTS[@]}"; do
>     [[ ${OCTET} -ge 0 ]] && [[ ${OCTET} -le 255 ]] || return 1
>   done
> }
> 
> # Checks whether an IPv4 network in CIRD notation is valid.
> #
> # @param ${1} = IPv4 network in CIDR notation
> function _is_ipv4_network_in_cidr_notation() {
>   local NETWORK ADDRESS SUFFIX
>   NETWORK=${1:?IPv4 address must be provided to _is_ipv4_network_in_cidr_notation}
> 
>   [[ ${NETWORK} == */* ]] || return 1
> 
>   ADDRESS=$(cut -d '/' -f 1 <<< "${NETWORK}")
>   _is_valid_ipv4_address "${ADDRESS}" || return 1
> 
>   # Lastly, check whether the suffix is valid
>   SUFFIX=$(cut -d '/' -f 2 <<< "${NETWORK}")
>   [[ ${SUFFIX} -ge 0 ]] && [[ ${SUFFIX} -le 32 ]] || return 1
> }
> 
> # Masks an individual octet (of an IP address).
> #
> # @param ${1} = prefix length
> # @param ${2} = octet
> # @output     = the masked octet
> function _mask_ip_octet() {
>   local PREFIX_LENGTH OCTET PREFIX_VALUES MASK
>   PREFIX_LENGTH=${1:?Prefix lentgh is required to _mask_ip_octet}
>   OCTET=${2:?Octet is required to _mask_ip_octet}
> 
>   if [[ ${PREFIX_LENGTH} -le 0 ]]; then
>     MASK=0
>   elif [[ ${PREFIX_LENGTH} -ge 8 ]]; then
>     MASK=255
>   else
>     PREFIX_VALUES=(0 128 192 224 240 248 252 254 255)
>     MASK=${PREFIX_VALUES[${PREFIX_LENGTH}]}
>   fi
> 
>   echo "$((OCTET &= MASK))"
> }
> 
> # Converts an IPv4 network in CIDR notation to the network's address.
> #
> # @param ${1} = the IPv4 network in CIDR notation
> # @output     = the network's address
> function _sanitize_ipv4_to_subnet_cidr() {
>   local NETWORK DIGIT_PREFIX_LENGTH MASKED_DIGITS OCTETS
> 
>   NETWORK=${1:?IPv4 network in CIDR notation must be provided to _sanitize_ipv4_to_subnet_cidr}
> 
>   if ! _is_ipv4_network_in_cidr_notation "${NETWORK}"; then
>     _log 'error' "${NETWORK} is not an IPv4 networm in CIDR notation"
>     return 1
>   fi
> 
>   DIGIT_PREFIX_LENGTH="${1#*/}"
>   _split_ipv4_into_octets OCTETS "${NETWORK%%/*}"
> 
>   for OCTET in "${OCTETS[@]}"; do
>     MASKED_DIGITS+=("$(_mask_ip_octet "${DIGIT_PREFIX_LENGTH}" "${OCTET}")")
>     DIGIT_PREFIX_LENGTH=$((DIGIT_PREFIX_LENGTH - 8))
>   done
> 
>   echo "${MASKED_DIGITS[0]}.${MASKED_DIGITS[1]}.${MASKED_DIGITS[2]}.${MASKED_DIGITS[3]}/${1#*/}"
> }
> ```
> 
> And the implementation of the new setup:
> 
> ```shell
> function _setup_local_networks() {
>   # for testing purposes
>   LOCAL_NETWORKS_IPV4="127.-1.0.1/8 GATEWAY 192.168.0.0/78 127.0.0.1/8 10.0.10.3/24"
> 
>   _log 'debug' "Adding local IPv4 networks '${LOCAL_NETWORKS_IPV4}' to services now"
> 
>   # Convert the string into an array we can work with
>   # Note: Do not remove the whitespace after `%s`; ref: https://stackoverflow.com/questions/12916352/shell-script-read-missing-last-line
>   mapfile -t -d ' ' NETWORKS < <(printf '%s ' "${LOCAL_NETWORKS_IPV4}")
> 
>   local NETWORK SANITIZED_NETWORKS
>   for NETWORK in "${NETWORKS[@]}"; do
>     _log 'trace' "Sanitizing waw network '${NETWORK}'"
> 
>     case "${NETWORK}" in
>       ( 'GATEWAY' )
>         NETWORK=$(ip route get default | /usr/bin/grep 'dev' | cut -d ' ' -f 4)
>         NETWORK=$(ip -4 addr show dev "${NETWORK}" | grep 'inet' | awk '{$1=$1;print}' | cut -d ' ' -f 2)
>         if [[ ${NETWORKS[*]} == *${NETWORK}* ]]; then
>           _log 'warning' "Gateway network '${NETWORK}' already listed in local networks ('${NETWORKS[*]}') - skipping"
>           continue
>         fi
>         _log 'trace' "Using gateway IP '${NETWORK}'"
>         ;;
> 
>       ( * ) ;;
>     esac
> 
>     if ! _is_ipv4_network_in_cidr_notation "${NETWORK}"; then
>       _log 'warning' "'${NETWORK}' is not a valid description of an IPv4 network in CIDR notation - skipping"
>       continue
>     fi
> 
>     SANITIZED_NETWORKS+=("${NETWORK}")
>   done
> 
>   # Update Postfix
>   postconf "mynetworks = ${SANITIZED_NETWORKS[*]}"
>   # Update Rspamd
>   echo "local_networks = \"${SANITIZED_NETWORKS[*]/%, }\";" >>"${RSPAMD_LOCAL_D}/options.inc"
> 
>   # Update OpenDKIM & OpenDMARC
>   for NETWORK in "${NETWORKS[@]}"; do
>     [[ ${ENABLE_OPENDKIM} -eq 1 ]] && echo "${NETWORK}" >>/etc/opendkim/TrustedHosts
>     [[ ${ENABLE_OPENDMARC} -eq 1 ]] && echo "${NETWORK}" >>/etc/opendmarc/ignore.hosts
>   done
> }
> 
> _setup_local_networks
> ```

ip route get default doesnt work for me: always returns local interface

**Comment by polarathene** [comment]
> I'd like to propose a new approach:
> 
> 1. Deprecate `PERMIT_DOCKER` and introduce `LOCAL_NETWORKS_IPV4` and `LOCAL_NETWORKS_IPV6`.
> 2. Use these ENV's values to write the appropriate configuration to all of our services (Postfix, Rspamd, OpenDKIM, OpenDMARC, you name it).

I don't have the bandwidth to go over the alternative implementation you've suggested right now, juggling too much and have a growing backlog.

While I agree that `PERMIT_DOCKER` is a terrible name and we need to change that at some point, the feature itself has value as the existing enum:

> ![image](https://github.com/docker-mailserver/docker-mailserver/assets/5098581/8d7b2aa4-919e-4670-ab25-78d70eedbde7)

It is nice that you can add trust for the implicit docker created networks, rather than be explicit. There's just the open relay caveat there if misconfigured.

- Supporting networks provided explicitly via user config is also worthwhile.
- I don't know what the IPv4 vs IPv6 compatibility is across services. I have suggested for Postfix to use the CIDR type table as an improvement as documented above.

Presently the primary issue being tracked here is that the current enum style config isn't implemented well. Those bugs should be fixed. Providing your own networks / IPs to trust is a separate feature and could be supplementary.

**Comment by georglauterbach** [comment]
Alright, I see. Actually, I think we want the same thing; my approach of `LOCAL_NETWORKS_IPV4` is basically a superset of the current `PERMIT_DOCKER`, i.e., when you previously used `PERMIT_DOCKER=container` but also have a network like `192.168.76.0/24` that you want to trust, you can use `LOCAL_NETWORKS_IPV4=container,192.168.76.0/24`.

The name would change (a breaking change of course), but the functionality would just be extended, not removed. What say ye?

**Comment by polarathene** [comment]
> The name would change (a breaking change of course), but the functionality would just be extended, not removed. What say ye?

If @casperklein agrees with the proposal I guess it's fine 👍 

- I kind of liked a single env managing the networks, not sure why it needs to split IPv4 / IPv6 (_assuming explained by code that I haven't looked over yet_).
- I'm also a bit iffy on mixing networks in with the enum switch.

I don't have any better input/alternatives to suggest, so if others agree in the value that will provide that's good enough for me :)

**Comment by polarathene** [comment]
Scheduled for v15, although you're welcome to fix the current support with `PERMIT_DOCKER` for v14, then follow-up for v15 with extending that to your proposal of `TRUSTED_NETWORKS` replacement which additionally supports providing a list of networks.

**Comment by georglauterbach** [comment]
Nah, v15 is what I would have gone for as well. I will not have time to implement an interim solution...
